modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/schema/TableDescriptorImpl.java

```diff
@@ -470,6 +470,31 @@ private IgniteCheckedException instantiationException(String typeName, Reflectiv
         return F.t(key, val);
     }
 
+    /** */
+    private <Row> IgniteBiTuple mergeTuple(Row row, List<String> updateColList, ExecutionContext<Row> ectx)
+        throws IgniteCheckedException {
+        RowHandler<Row> hnd = ectx.rowHandler();
+
+        int rowColumnsCnt = hnd.columnCount(row);
+
+        if (rowColumnsCnt == descriptors.length)
+            return F.t(insertTuple(row, ectx), null); // Only WHEN NOT MATCHED clause in MERGE.
+        else if (rowColumnsCnt == descriptors.length + updateColList.size())
+            return F.t(null, updateTuple(row, updateColList, 0, ectx)); // Only WHEN MATCHED clause in MERGE.
+        else {
+            // Both WHEN MATCHED and WHEN NOT MATCHED clauses in MERGE.
+            assert rowColumnsCnt == descriptors.length * 2 + updateColList.size() : "Unexpected columns count: " +
+                rowColumnsCnt;
+
+            int updateOffset = descriptors.length; // Offset of fields for update statement.
+
+            if (hnd.get(updateOffset + QueryUtils.KEY_COL, row) != null)
+                return F.t(null, updateTuple(row, updateColList, updateOffset, ectx));
```

Comment:
Honestly speaking, tuple in tuple freaks me out. I would prefer to introduce a more clear abstraction for modify rows. Besides, it would help to ease the support  of `WHEN MATCHED THEN DELETE` clause.

```
<merge when matched clause> ::=
  WHEN MATCHED [ AND <search condition> ]
      THEN <merge update or delete specification>

<merge update or delete specification> ::=
    <merge update specification>
  | <merge delete specification>
...
<merge delete specification> ::=
  DELETE
```

Reply:
Ok, I will change T2 to T3 with the operation type as the third field.
BTW, what standard supports merge delete? In 2003 standard only merge update is supported. Currently, Calcite also supports only merge update.

Reply:
Would it be better to replace tuples with a POJO? Something like ModifyTuple(key, value, op)? 

> what standard supports merge delete

I looked at 2011 standard. It appears this feature were introduced in 2011 standard.

Reply:
I mean tuple is great utility containers, but, personally, I would prefer to avoid their using in API, even internal one.

Reply:
Fixed